part-1.py

```diff
@@ -5,16 +5,34 @@
 
 # factorial
 
+def factorial(num):
+    if num < 0:
+        raise ValueError
+    if num == 1 or num == 0:
+        return 1
+    return num * factorial(num - 1)
 
 
 # reverse
 
-
+def reverse(text):
```

Comment:
👍 , just note that `text[0:-1]` makes a copy of the string element-by-element.  That causes this to be O(n^2) for time/space complexity.